### PR TITLE
Refactor MethodBind to use variadic templates

### PR DESCRIPTION
**Pull request by reduz**
Removed make_binders and the old style generated binders.

-Use more modern C++
-Much less code and easier to maintain
-Hopefully faster compile times.

May have disadvantages:

-Build size might increase in some platforms. Needs to be tested

*Bugsquad edit: Follow-up to #42780.*

### CODE REVIEW

File: core/binder_common.h [hunk 1]
```diff
@@ -0,0 +1,650 @@
+#ifndef BINDER_COMMON_H
```
Comment (akien-mga) [inline core/binder_common.h:1]: Missing header.

**Comment by reduz** [comment]
Any way to check build sizes in the CI ?

**Comment by Calinou** [comment]
@reduz Adding a `ls -l bin` command at the CI script should do the trick, I guess. Just make sure to remove it before the PR is merged.

**Comment by akien-mga** [comment]
Note: I get many warnings with GCC 10.2.0 that the CI somehow doesn't catch, I'll fix them before this is merged.

File: core/class_db.h [hunk 2]
```diff
@@ -36,7 +36,7 @@
 #include "core/print_string.h"
 
 /** To bind more then 6 parameters include this:
- *  #include "core/method_bind_ext.gen.inc"
+ *
  */
```
Comment (Xrayez) [inline core/class_db.h:40]: Does it mean that it's not required to include `core/method_bind_ext.gen.inc` to have more than 5 parameters while binding methods anymore? See godotengine/godot-proposals#1229. If yes, then there are several places where this limit is advertised in the documentation as well, such as [Custom modules](https://docs.godotengine.org/en/latest/development/cpp/custom_modules_in_cpp.html).

**Comment by akien-mga** [comment]
Force pushed to fix GCC warnings I have locally with `warnings=extra`.

**Comment by akien-mga** [comment]
Some size comparisons:

- Commit 88a3db5bff3c5d62aec7e7ade4f6c62bf2d0ec3e:

```
-rwxr-xr-x 1 runner docker 281393592 Oct 15 20:39 godot.linuxbsd.opt.tools.64.mono
-rwxr-xr-x 1 runner docker 148871320 Oct 15 20:38 godot.linuxbsd.opt.64.mono
-rwxr-xr-x  1 runner  staff  132428004 Oct 15 20:41 godot.osx.opt.tools.64
-rwxr-xr-x  1 runner  staff  63961008 Oct 15 20:38 godot.osx.opt.64
-rw-r--r--  1 runner  staff  68400448 Oct 15 20:44 libgodot.iphone.opt.arm64.a
-a---          10/15/2020  8:42 PM       81804800 godot.windows.opt.tools.64.exe
-a---          10/15/2020  8:43 PM       33187328 godot.windows.opt.64.exe
```

- This PR (commit e5aa1f06059f037850c54eb44b6aef99c2c931e8)

```
-rwxr-xr-x 1 runner docker 156715344 Oct 16 13:03 godot.linuxbsd.opt.64.mono
-rwxr-xr-x  1 runner  staff  68010624 Oct 16 13:15 godot.osx.opt.64
 -rw-r--r--  1 runner  staff  70698864 Oct 16 13:16 libgodot.iphone.opt.arm64.a
-a---          10/16/2020  1:10 PM       30804992 godot.windows.opt.64.exe
```

So impact on size:
- Linux GCC template: **+5.3%**
- Linux GCC editor: TODO
- macOS Clang template: **+6.3%**
- macOS Clang editor: TODO
- iOS Clang template: **+3.4%**
- Windows MSVC template: **-7.2%**
- Windows MSVC editor: TODO
- Android: don't know, `ls -l bin` doesn't work for it as the lib is in `platform/android`

**Comment by reduz** [comment]
Force pushed a new version using the UNTYPED_METHOD_BIND hack for compilers where it works, lets see if there is any size reduction.

**Comment by akien-mga** [comment]
For comparison with previous commit:

- This PR (commit 878971c58ffcf1f9215e3367e5951fa9b13a5773)

```
-rwxr-xr-x 1 runner docker 281331576 Oct 16 19:51 godot.linuxbsd.opt.tools.64.mono
-rwxr-xr-x 1 runner docker 148084944 Oct 16 19:42 godot.linuxbsd.opt.64.mono
-rwxr-xr-x  1 runner  staff  137931308 Oct 16 19:51 godot.osx.opt.tools.64
-rwxr-xr-x  1 runner  staff  69141440 Oct 16 19:42 godot.osx.opt.64
-rw-r--r--  1 runner  staff  69569160 Oct 16 19:39 libgodot.iphone.opt.arm64.a
-a---          10/16/2020  7:59 PM       78453760 godot.windows.opt.tools.64.exe
-a---          10/16/2020  7:45 PM       31036416 godot.windows.opt.64.exe
```

Impact on size (compared with `master`):

- Linux GCC template: **-0.5%** (good)
- Linux GCC editor: **-0.0%** (good)
- macOS Clang template: **+8.1%** (worse)
- macOS Clang editor: **+4.2%** (worse)
- iOS Clang template: **+1.7%** (good)
- Windows MSVC template: **-6.5%** (good)
- Windows MSVC editor: **-4.1%** (good)

Optimized local build: `scons p=linuxbsd target=release tools=no use_lto=yes debug_symbols=no`

```
-rwxr-xr-x 1 akien akien  48004968 Oct 16 21:39 godot.linuxbsd.opt.64.before*
-rwxr-xr-x 1 akien akien  51585104 Oct 16 21:57 godot.linuxbsd.opt.64.pr42826.e5aa1f0*
-rwxr-xr-x 1 akien akien  46447360 Oct 16 21:57 godot.linuxbsd.opt.64.pr42826.878971c*
```
So size reduction by **-3.2%** for Linux/GCC.

Conclusion:
- Linux GCC: 878971c is much better, same size as without the PR, or even 3% when optimized
- macOS Clang: 878971c makes things a bit worse :|
- iOS Clang: 878971c improves things a bit, the size increase is OK
- Windows MSVC: Both commits are great, 878971c gives slightly less size reduction, but it's still very good

File: core/class_db.h (hunk 2 above)
Comment (reduz) [reply core/class_db.h:40]: No. This is no longer required as everything uses variadic templates. Feel free to PR changes to the docs.

**Comment by reduz** [comment]
This is complete from my part. Bindings should be smaller in all platforms.

**Comment by akien-mga** [comment]
Size comparisons:

- Reference (#42861):
```
-rwxr-xr-x 1 runner docker 110443760 Oct 16 20:59 godot.linuxbsd.opt.tools.64.mono
-rwxr-xr-x 1 runner docker 51314856 Oct 16 20:41 godot.linuxbsd.opt.64.mono
-rwxr-xr-x  1 runner  staff  113471492 Oct 16 20:54 godot.osx.opt.tools.64
-rwxr-xr-x  1 runner  staff  51788272 Oct 16 20:41 godot.osx.opt.64
-rw-r--r--  1 runner  staff  68400448 Oct 16 20:26 libgodot.iphone.opt.arm64.a
-a---          10/16/2020  8:56 PM       81803264 godot.windows.opt.tools.64.exe
-a---          10/16/2020  8:40 PM       33186816 godot.windows.opt.64.exe
```

- This PR (commit 3d27d3e):
```
-rwxr-xr-x 1 runner docker 109229304 Oct 17 18:22 godot.linuxbsd.opt.tools.64.mono
-rwxr-xr-x 1 runner docker 50452296 Oct 17 18:21 godot.linuxbsd.opt.64.mono
-rwxr-xr-x  1 runner  staff  109035820 Oct 17 18:22 godot.osx.opt.tools.64
-rwxr-xr-x  1 runner  staff  49164408 Oct 17 18:22 godot.osx.opt.64
-rw-r--r--  1 runner  staff  68937488 Oct 17 18:21 libgodot.iphone.opt.arm64.a
-a---          10/17/2020  6:24 PM       78444544 godot.windows.opt.tools.64.exe
-a---          10/17/2020  6:23 PM       30988288 godot.windows.opt.64.exe
```

- Linux GCC editor: **-1.1%**
- Linux GCC template: **-1.7%**
- macOS Clang editor: **-3.9%**
- macOS Clang template: **-5.1%**
- iOS Clang template: **+0.8%**
- Windows MSVC editor:  **-4.1%**
- Windows MSVC template: **-6.6%**

All pretty good :)

**Comment by akien-mga** [comment]
For the reference, the LinuxBSD and Server platforms (the latter is dead code though) still have code to enable `TYPED_METHOD_BIND` for LLVM, as was used by macOS (which caused the size increase in previous attempts).

MSVC and UWP also use `TYPED_METHOD_BIND`.

**Review by akien-mga — APPROVED** [review]
Tested locally with GCC 10.2 and it builds without warnings, and runs fine.
Code style is good, I pushed some last nitpicks as a last amend.

I undid the CI changes, though I might do them in a dedicated PR as I think it makes sense for artifacts to be without debug symbols (much smaller).

See my last comments for a comparison of the size change from this PR. Pretty positive overall.

**Comment by akien-mga** [comment]
I compiled for Javascript locally to check the sizes:
```
$ ll bin/js-*
bin/js-master:
total 17368
-rw-r--r-- 1 akien akien    12566 Oct 18 14:17 godot.javascript.opt.engine.js
-rw-r--r-- 1 akien akien   206987 Oct 18 14:22 godot.javascript.opt.js
-rw-r--r-- 1 akien akien 13147718 Oct 18 14:22 godot.javascript.opt.wasm
-rw-r--r-- 1 akien akien   219554 Oct 18 14:22 godot.javascript.opt.wrapped.js
-rw-r--r-- 1 akien akien  4184508 Oct 18 14:22 godot.javascript.opt.zip

bin/js-pr42826:
total 17204
-rw-r--r-- 1 akien akien    12566 Oct 18 13:55 godot.javascript.opt.engine.js
-rw-r--r-- 1 akien akien   206987 Oct 18 13:55 godot.javascript.opt.js
-rw-r--r-- 1 akien akien 13014435 Oct 18 13:55 godot.javascript.opt.wasm
-rw-r--r-- 1 akien akien   219554 Oct 18 13:55 godot.javascript.opt.wrapped.js
-rw-r--r-- 1 akien akien  4147597 Oct 18 13:55 godot.javascript.opt.zip
```

So **-1.0%** for the `.wasm` file, which gives **-0.9%** on the .zip (37 kB saved).

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**